HOC.md

```diff
@@ -0,0 +1,153 @@
+# HOC Cheatsheet
+
+**This HOC Cheatsheet** compiles all available knowledge for writing Higher Order Components with React and TypeScript.
+
+- We will map closely to [the official docs on HOCs](https://reactjs.org/docs/higher-order-components.html) initially
+- While hooks exist, many libraries and codebases still have a need to type HOCs.
+- Render props may be considered in future
+- The goal is to write HOCs that offer type safety while not getting in the way.
+
+---
+
+### HOC Cheatsheet Table of Contents
+
+<details>
+
+<summary><b>Expand Table of Contents</b></summary>
+
+- [Section 0: Prerequisites](#section-0-prerequisites)
+  </details>
+
+# Section 1: React HOC docs in TypeScript
+
+In this first section we refer closely to [the React docs on HOCs](https://reactjs.org/docs/higher-order-components.html) and offer direct TypeScript parallels.
+
+## Example: Use HOCs For Cross-Cutting Concerns
+
+<details>
+
+<summary>
+<b>Misc variables reference in the example below</b>
+</summary>
+
+```tsx
+/** dummy child components that take anything */
+const Comment = (_: any) => null;
+const TextBlock = Comment;
+
+/** dummy Data */
+type CommentType = { text: string; id: number };
+const comments: CommentType[] = [
+  {
+    text: 'comment1',
+    id: 1
+  },
+  {
+    text: 'comment2',
+    id: 2
+  }
+];
+const blog = 'blogpost';
+
+/** mock data source */
+const DataSource = {
+  addChangeListener(e: Function) {
+    // do something
+  },
+  removeChangeListener(e: Function) {
+    // do something
+  },
+  getComments() {
+    return comments;
+  },
+  getBlogPost(id: number) {
+    return blog;
+  }
+};
+/** type aliases just to deduplicate */
+type DataType = typeof DataSource;
+type TODO_ANY = any;
+
+/** utility types we use */
+type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
+type Optionalize<T extends K, K> = Omit<T, keyof K>;
+
+/** Rewritten Components from the React docs that just uses injected data prop */
+function CommentList({ data }: WithSubscriptionProps) {
+  return (
+    <div>
+      {data.map((comment: CommentType) => (
+        <Comment comment={comment} key={comment.id} />
+      ))}
+    </div>
+  );
+}
+interface BlogPostProps extends WithSubscriptionProps {
+  id: number;
+  // children: ReactNode;
+}
+function BlogPost({ data, id }: BlogPostProps) {
+  return (
+    <div>
+      <TextBlock text={data} />;
+    </div>
+  );
+}
+```
+
+</details>
+
+```tsx
+// // ACTUAL HOC
+interface WithSubscriptionProps {
+  data: TODO_ANY;
+}
+/** This function takes a component... */
+function withSubscription<
+  T extends WithSubscriptionProps = WithSubscriptionProps
+>(
+  WrappedComponent: React.ComponentType<T>,
+  selectData: (DataSource: DataType, props: TODO_ANY) => TODO_ANY
```

Comment:
I really don't like this line but i cant figure out how to type it. Maybe with an `unknown` for the return value? but I feel i can infer stuff too. cc @ferdaber

Reply:
Here's a working example-- you can remove and add stuff as needed:
```tsx
import React, { Component, JSXElementConstructor } from 'react'

// boilerplate
declare const DataSource: {
  addChangeListener(handleChange: () => void): void
  removeChangeListener(handleChange: () => void): void
}

// data is generic
interface WithDataProps<T> {
  data: T
}

// T is the type of data
// P is the props of the wrapped component that is inferred
// C is the actual interface of the wrapped component (used to grab defaultProps from it)
export function withSubscription<T, P extends WithDataProps<T>, C>(
  // this type allows us to infer P, but grab the type of WrappedComponent separately without it interfering with the inference of P
  WrappedComponent: JSXElementConstructor<P> & C,
  // selectData is a functor for T
  // props is Readonly because it's readonly inside of the class
  selectData: (
    dataSource: typeof DataSource,
    props: Readonly<JSX.LibraryManagedAttributes<C, Omit<P, 'data'>>>
  ) => T
) {
  // the magic is here: JSX.LibraryManagedAttributes will take the type of WrapedComponent and resolve its default props
  // against the props of WithData, which is just the original P type with 'data' removed from its requirements
  type Props = JSX.LibraryManagedAttributes<C, Omit<P, 'data'>>
  type State = {
    data: T
  }
  class WithData extends Component<Props, State> {
    constructor(props: Props) {
      super(props)
      this.handleChange = this.handleChange.bind(this)
      this.state = {
        data: selectData(DataSource, props),
      }
    }

    componentDidMount() {
      DataSource.addChangeListener(this.handleChange)
    }

    componentWillUnmount() {
      DataSource.removeChangeListener(this.handleChange)
    }

    handleChange() {
      this.setState({
        data: selectData(DataSource, this.props),
      })
    }

    render() {
      // the typing for spreading this.props is... very complex. best way right now is to just type it as any
      // data will still be typechecked
      return <WrappedComponent data={this.state.data} {...this.props as any} />
    }
  }
  return WithData
}
```

Reply:
got it. needed to finagle a bit of stuff to achieve my ideal "spoonfed" demo but got it working. cheers bud